azurerm/resource_arm_private_dns_mx_record.go

```diff
@@ -0,0 +1,235 @@
+		return fmt.Errorf("Error creating/updating Private DNS MX Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
+	}
+
+	resp, err := client.Get(ctx, resGroup, zoneName, privatedns.MX, name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Private DNS MX Record %q (Zone %q / Resource Group %q): %s", name, zoneName, resGroup, err)
+	}
+
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Private DNS MX Record %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceArmPrivateDnsMxRecordRead(d, meta)
+}
+
+func resourceArmPrivateDnsMxRecordRead(d *schema.ResourceData, meta interface{}) error {
+	dnsClient := meta.(*ArmClient).PrivateDns.RecordSetsClient
+	ctx, cancel := timeouts.ForRead(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["MX"]
+	zoneName := id.Path["privateDnsZones"]
+
+	resp, err := dnsClient.Get(ctx, resGroup, zoneName, privatedns.MX, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading Private DNS MX record %s: %+v", name, err)
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("zone_name", zoneName)
+	d.Set("ttl", resp.TTL)
+
+	if err := d.Set("record", flattenAzureRmPrivateDnsMxRecords(resp.MxRecords)); err != nil {
+		return err
+	}
+
+	return tags.FlattenAndSet(d, resp.Metadata)
+}
+
+func resourceArmPrivateDnsMxRecordDelete(d *schema.ResourceData, meta interface{}) error {
+	dnsClient := meta.(*ArmClient).PrivateDns.RecordSetsClient
+	ctx, cancel := timeouts.ForDelete(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["MX"]
+	zoneName := id.Path["privateDnsZones"]
+
+	_, err = dnsClient.Delete(ctx, resGroup, zoneName, privatedns.MX, name, "")
+	if err != nil {
+		return fmt.Errorf("Error deleting Private DNS MX Record %s: %+v", name, err)
+	}
+
+	return nil
+}
+
+func flattenAzureRmPrivateDnsMxRecords(records *[]privatedns.MxRecord) []map[string]interface{} {
+	results := make([]map[string]interface{}, 0)
+
+	if records != nil {
+		for _, record := range *records {
+			if record.Preference == nil ||
+				record.Exchange == nil {
+				continue
+			}
+
+			results = append(results, map[string]interface{}{
+				"preference": *record.Preference,
+				"exchange":   *record.Exchange,
+			})
+		}
+	}
+
+	return results
+}
+
+func expandAzureRmPrivateDnsMxRecords(d *schema.ResourceData) *[]privatedns.MxRecord {
+	recordStrings := d.Get("record").(*schema.Set).List()
+	records := make([]privatedns.MxRecord, len(recordStrings))
+
+	for i, v := range recordStrings {
+		record := v.(map[string]interface{})
```

Comment:
`v` could be nil and we should confirm that it's not before casting it to a `map[string]interface{}`